src/line_launcher/mod.rs

```diff
@@ -20,6 +20,44 @@ use playing_state::PlayingState;
 mod progression_state;
 use progression_state::ProgressionState;
 
+enum DurationBetweenSixteenthNotes {
+    Uninitialized,
+    PartiallyInitialized {
+        last_timestamp: SystemTime,
+    },
+    Initialized {
+        last_timestamp: SystemTime,
+        last_duration: Duration,
+    },
+}
+
+impl DurationBetweenSixteenthNotes {
+    pub fn process_beat_message(&mut self, _beat_message: &BeatNumber) {
+        let now = SystemTime::now();
+        *self = match *self {
+            DurationBetweenSixteenthNotes::Uninitialized => {
+                DurationBetweenSixteenthNotes::PartiallyInitialized {
+                    last_timestamp: now,
+                }
+            }
+            DurationBetweenSixteenthNotes::PartiallyInitialized { last_timestamp }
+            | DurationBetweenSixteenthNotes::Initialized { last_timestamp, .. } => {
+                DurationBetweenSixteenthNotes::Initialized {
+                    last_timestamp: now,
+                    last_duration: now.duration_since(last_timestamp).unwrap(),
```

Comment:
Same comment as below: I'd deref `last_timestamp` rather than `self` (in `match *self).

Reply:
Ok updated to only deref `last_timestamp`